### PR TITLE
fix(Scripts/ZulGurub: Mandokir's Threatening Gaze.

### LINKED ISSUE #11748: [ZG] Bloodlord Mandokir - Threatening Gaze
### Current Behaviour

Boss casts Threatening Gaze.

~~The debuff has no effect. Nothing happens.~~ fixed by #11854

 - Boss cannot attack while casting.

 - Upon an action the boss charges immediately rather than after the debuff runs out.

 - ~~Charge movement is wonky. Player can move while the boss is charging.~~ Blizzlike ([proof](https://www.youtube.com/watch?v=28EhZvGtf0s))

 - If player is more than 10y away from the boss, Guillotine is not cast.

 - Some spells that do not generate threat are triggering the charge, like using Hearthstone.

Edit, new bugs:

 - Auto-attacks aren't triggering the charge.

 - If the Charge is triggered while the boss is casting Whirlwind, the cast is not cancelled. The charge effects occur but he does not move. I don't know the blizzlike method but this doesn't seem right.

### Expected Blizzlike Behaviour

Boss should cast Threatening Gaze at a random player.

Boss can attack while casting Threatening Gaze.

**After the debuff runs out**, if the player affected did any actions that generate threat and are not movement, HoT ticks or DoT ticks, Mandokir will cast a Charge spell followed by Guillotine, and then return to the main target. I do not know what effects spells such as Feign Death and Vanish have in this mechanic.

Threatening Gaze can be broken by Line of Sight or if there's not a clear path to the target.

### Source

https://user-images.githubusercontent.com/100873791/168453776-9f1a40c1-50e6-4a7d-9c70-10ab3eea9744.mp4

Boss can attack while casting: https://youtu.be/WyaERCgkJBU?t=66

Threatening Gaze Charge can be LoS'd or avoided:
https://youtu.be/PLHmWYmHieg?t=133
https://youtu.be/yQIhi37rlXM?t=69 afterwards the Threatening Gaze does not affect the player
Many times throughout https://youtu.be/R7oje40mu0Q?t=1389 
https://youtu.be/FcFWZJLlLZA?t=52
https://www.youtube.com/watch?v=ltXE4QGeYpU this strategy was famous for a reason, same thing with the 'spear wall' strategy that was broken in following patches after the maps changed.

Threatening Gaze:
[From Wowhead comments](https://tbc.wowhead.com/npc=11382/bloodlord-mandokir#comments:id=2826633):
>If you are being watched you must stop anything that causes aggro; meaning: Healing, debuffing, buffing, potion drinking, bandaging, dealing damage and cleasing. You may move, you may jump, you may pull additional adds (not recommended though) and I think you even may heartstone out.

DoT ticks, HoT ticks and Totems ticks do not count as an action: https://youtu.be/QsC_Xwas01s?t=30

### Steps to reproduce the problem

1. `.go xyz -12212 -1959 133 309`
2. Kill the Vilebranch Speaker and start the encounter.

### Extra Notes

Bloodlord Mandokir (ID: 11382)
Threating Gaze (ID: 24314)
Threating Gaze - Charge (ID: 24315)
Threating Gaze - Guillotine (ID: 24316)

From what I've observed in several videos, the timers seem to be 12-24s for Threatening Gaze.

If the player uses `.learn 24315` and casts it, it seems that it deals damage if the target is close enough.

### AC rev. hash/commit

https://github.com/azerothcore/azerothcore-wotlk/commit/5474b00a3d4f75245068535c26fa44c70f7ca6b4

### Operating system

Windows 10

### Custom changes or Modules

_No response_

### PR DESCRIPTION
**Pull request by UltraNix**
Boss can melee attack while casting.
Boss charges after the debuff run out.
Guillotine is casted on hit rather than on launch.
Properly coded Threatening Gaze Charge (with dot/hot exceptions)
Whirlwind is removed during charge.
Fixes #11748

## Issues Addressed:

- Closes #11748

## Tests Performed:

- Not tested.

## How to Test the Changes:

`.go xyz -12212 -1959 133 309`
Kill the Vilebranch Speaker and start the encounter.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Gultask** [comment]
Nef was already working on it https://github.com/azerothcore/azerothcore-wotlk/pull/12072
But your PR fixes stuff that was on Nef's to-do list:

- Boss can melee attack while casting.
- Guillotine is casted on hit rather than on launch.
- Whirlwind is removed during charge.
- DoT exceptions

**Comment by UltraNix** [comment]
This PR superseded Nefertum's part of Threatening Gaze.

**Comment by Gultask** [comment]
ok. I'll test it asap :)

**Comment by acidmanifesto** [comment]
Just posting this here for faster linking.
https://github.com/azerothcore/azerothcore-wotlk/pull/12072

@UltraNix will this conflict with his pr in anyway? just asking so i will know what needs to be merged in what order.

**Comment by UltraNix** [comment]
Yes - will confict.

**Comment by Gultask** [comment]
- [ ] Healing didn't trigger a charge.

 - [ ] I've not seen Guillotine being cast, regardless if I was close to the boss or not.

 - [ ] Boss is not attacking while casting Threatening Charge.

But everything else is working from what I saw.

Video of him not attacking while casting (at 5s):

https://user-images.githubusercontent.com/100873791/174444911-5a650292-48d5-4b28-ad80-10da370c1e2c.mp4

**Review by acidmanifesto — APPROVED** [review]

**Comment by acidmanifesto** [comment]
> Yes - will confict.

ill approve the run and let you two discuss the approach for intergrating this into azerothcore.

**Comment by acidmanifesto** [comment]
@Gultask thank you for the test results.

**Comment by UltraNix** [comment]
@Gultask 
`Healing didn't trigger a charge. (Lesser Heal, for example)`
Does healing somebody increases threat done to boss?

**Comment by Gultask** [comment]
I healed myself, and it did increase threat by 1 or 2 since it was a rank 1 heal. I'll try again with a Greater Heal and see if there's a difference.

**Comment by UltraNix** [comment]
@Gultask Please, retest it.

**Comment by Gultask** [comment]
I, uh... got build errors.
I'll reclone my local so we hopefully don't get a repeat of that Renataki thing :p 

image

**Comment by UltraNix** [comment]
@Gultask Try now.

**Comment by Gultask** [comment]
Everything seems to be working from what I can see! 👍

File: src/server/game/Combat/ThreatMgr.cpp [hunk 1]
```diff
@@ -32,11 +32,11 @@
 //==============================================================
 
 // The hatingUnit is not used yet
-float ThreatCalcHelper::calcThreat(Unit* hatedUnit, Unit* /*hatingUnit*/, float threat, SpellSchoolMask schoolMask, SpellInfo const* threatSpell)
+float ThreatCalcHelper::calcThreat(Unit* hatedUnit, Unit* hatingUnit, float threat, SpellSchoolMask schoolMask, SpellInfo const* threatSpell)
```
Comment (acidmanifesto) [inline src/server/game/Combat/ThreatMgr.cpp:35]: unused variable
image

Comment (acidmanifesto) [reply src/server/game/Combat/ThreatMgr.cpp:35]:
```
float ThreatCalcHelper::calcThreat(Unit* hatedUnit, Unit* /*hatingUnit*/, float threat, SpellSchoolMask schoolMask, SpellInfo const* threatSpell)
```

**Comment by Nefertumm** [comment]
There are still build issues

**Comment by Nefertumm** [comment]
image

**Comment by LannyE** [comment]
I just updated my wowemu this morning and got this:
image

The last time I updated was June 22. These errors span across three different PRs. #12095 #12134 #12200 

I checked the cpps that added these and they are updated. Not sure what the problem is.

**Comment by acidmanifesto** [comment]
image
I get no such errors on current revision [b866c47](https://github.com/azerothcore/azerothcore-wotlk/commit/b866c470d63644ad09f5b7d469f52dd9eb78303a)

**Comment by LannyE** [comment]
Then it has got to be something I did.  I did another complete new pull and build and the errors went away.  Sorry about that.  Thanks for the help.

**Pull request merged**